ENG-3712 Implement getComponentVersion

raxkaynan opened the pull request:
Implements the method by extracting the actual version of a component; Refactors ComponentService

raxkaynan committed:
ENG-3712 Implement getComponentVersion

Implements the method by extracting the actual version of a component; Refactors ComponentService

cecchisandrone commented on src/models/component.ts:
```diff
@@ -1,11 +1,21 @@
+  stack: string
+}
+
+export interface Component extends PartialComponent, Record<string, unknown> {
```
I suggest to keep `Component` as a clean type mapping what we have in the descriptor. If you need smth custom for visualization (table) better extending from it and creating a `VersionedComponent`

cecchisandrone commented on src/services/component-descriptor-service.ts:
```diff
@@ -1,9 +1,43 @@
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  getComponentVersion(name: string, type: ComponentType): string {
-    return '0.0.1'
+  public getComponentVersion(
```
Make this function take only `Component` as parameter

cecchisandrone commented on src/services/component-descriptor-service.ts:
```diff
@@ -1,9 +1,43 @@
+        this.parsePomXML(compPath).project?.version?.[0]
+
+    if (!version) {
+      throw new CLIError(`Failed to get version of ${name} ${type}`)
```
I would not throw error in case of missing version. Better returning `unkown` for the moment

cecchisandrone commented on src/services/component-descriptor-service.ts:
```diff
@@ -1,9 +1,43 @@
+
+  private parsePackageJSON(dir: string): { version: string } {
+    return JSON.parse(
+      fs.readFileSync(path.resolve(dir, 'package.json'), 'utf-8')
```
Handle exception if file is not found/readable. In such case we fallback to `unknown` for the moment

cecchisandrone commented on src/services/component-descriptor-service.ts:
```diff
@@ -1,9 +1,43 @@
+  private parsePomXML(dir: string): Partial<{ project: { version: string } }> {
+    let result: Partial<{ project: { version: string } }> = {}
+    xml2js.parseString(
+      fs.readFileSync(path.resolve(dir, 'pom.xml'), 'utf-8'),
```
Same as above

raxkaynan committed:
ENG-3712 Refactor and amend List, ComponentDescriptorService, ComponentService

Replace PartialComponent with Component and add VersionedComponent; allow undefined value for component version
